akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala

```diff
@@ -487,61 +471,47 @@ private[persistence] trait Eventsourced extends Snapshotter with Stash with Stas
+        changeState(initializing(recoveryBehavior))
+        journal ! ReadHighestSequenceNr(lastSequenceNr, persistenceId, self)
+
         onReplayFailure(cause) // callback for subclass implementation
-        // FIXME what happens if RecoveryFailure is handled, i.e. actor is not stopped?
-        Eventsourced.super.aroundReceive(recoveryBehavior, RecoveryFailure(cause))
+        Eventsourced.super.aroundReceive(recoveryBehavior, RecoveryFailure(cause)(None))
       case other ⇒
         internalStash.stash()
     }
   }
 
   /**
-   * Consumes remaining replayed messages and then changes to `prepareRestart`. The
-   * message that caused the exception during replay, is re-added to the mailbox and
-   * re-received in `prepareRestart` state.
+   * Consumes remaining replayed messages and then emits RecoveryFailure to the
+   * `receiveRecover` behavior.
    */
-  private def replayFailed(cause: Throwable, failureMessage: Envelope) = new State {
+  private def replayFailed(recoveryBehavior: Receive, cause: Throwable, failed: PersistentRepr) = new State {
 
     override def toString: String = "replay failed"
     override def recoveryRunning: Boolean = true
 
     override def stateReceive(receive: Receive, message: Any) = message match {
-      case ReplayMessagesFailure(_) ⇒
-        replayCompleted()
-        // journal couldn't tell the maximum stored sequence number, hence the next
-        // replay must be a full replay (up to the highest stored sequence number)
-        // Recover(lastSequenceNr) is sent by preRestart
-        setLastSequenceNr(Long.MaxValue)
-      case ReplayMessagesSuccess ⇒ replayCompleted()
-      case ReplayedMessage(p)    ⇒ updateLastSequenceNr(p)
-      case r: Recover            ⇒ // ignore
-      case _                     ⇒ internalStash.stash()
+      case ReplayedMessage(p) ⇒ updateLastSequenceNr(p)
+      case ReplayMessagesSuccess | ReplayMessagesFailure(_) ⇒ replayCompleted()
+      case r: Recover ⇒ // ignore
+      case _ ⇒ internalStash.stash()
     }
 
     def replayCompleted(): Unit = {
-      changeState(prepareRestart(cause))
-      mailbox.enqueueFirst(self, failureMessage)
-    }
-  }
-
-  /**
-   * Re-receives the replayed message that caused an exception and re-throws that exception.
-   */
-  private def prepareRestart(cause: Throwable) = new State {
-    override def toString: String = "prepare restart"
-    override def recoveryRunning: Boolean = true
+      // in case the actor resumes the state must be initializing
+      changeState(initializing(recoveryBehavior))
+      journal ! ReadHighestSequenceNr(failed.sequenceNr, persistenceId, self)
 
-    override def stateReceive(receive: Receive, message: Any) = message match {
-      case ReplayedMessage(_) ⇒ throw cause
-      case _                  ⇒ // ignore
+      Eventsourced.super.aroundReceive(recoveryBehavior,
+        RecoveryFailure(cause)(Some((failed.sequenceNr, failed.payload))))
```

Comment:
Nope, I see, if unhandled it will throw. So handling a RecoveryFailure now means that it must actually be handled, right?